File: dbusmenu4j/src/main/java/io/gitlab/jfronny/dbusmenu4j/Peer.java
package io.gitlab.jfronny.dbusmenu4j;

import io.gitlab.jfronny.commons.unsafe.reflect.Reflect;
import io.gitlab.jfronny.commons.unsafe.reflect.impl.CoreReflect;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.awt.*;
import java.lang.invoke.MethodHandle;
import java.lang.reflect.*;
import java.util.function.*;

public sealed interface Peer {
    long nativePointer();

    static Peer coerce(Object peer) {
        if (peer.getClass().getName().contains("X11")) return new X11(peer);
        if (peer.getClass().getName().contains("WL")) return new WL(peer);
        throw new IllegalArgumentException("Unsupported peer type: " + peer.getClass());
    }

    final class X11 implements Peer {
        private static final Class<?> klazz;
        private static final Method getPtrMethod;

        static {
            try {
                klazz = Class.forName("sun.awt.X11.XBaseWindow");
                getPtrMethod = klazz.getDeclaredMethod("getWindow");
            } catch (ClassNotFoundException | NoSuchMethodException e) {
                Resolver.logger.error("Failed to resolve X11 peer methods", e);
                throw new RuntimeException(e);
            }
        }

        private final Object peer;

        public X11(Object peer) {
            this.peer = klazz.cast(peer);
        }

        @Override
        public long nativePointer() {
            try {
                return (Long) getPtrMethod.invoke(peer);
            } catch (IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        }
    }

    final class WL implements Peer {
        private static final Class<?> klazz;
        private static final Class<?> containerPeerClass;
        private static final Function<Object, Insets> getInsetsMethod;
        private static final Field nativePtrField;
        private static final BiConsumer<Object, Runnable> performLockedMethod;
        private static final Field wlSurfaceField;
        private static final Field decorationField;
        private static final Consumer<Object> markRepaintNeededMethod;
        private static final Field isUndecoratedField;
        private static final Function<Object, Long> getWlSurfacePtr;

        private final Object peer;

        static {
            try {
                klazz = Class.forName("sun.awt.wl.WLComponentPeer");
                nativePtrField = Resolver.withAccess(klazz.getDeclaredField("nativePtr"));
                performLockedMethod = Resolver.uncheck(Reflect.instanceProcedure(klazz, "performLocked", Runnable.class));
                wlSurfaceField = Resolver.withAccess(klazz.getDeclaredField("wlSurface"));
                containerPeerClass = Class.forName("java.awt.peer.ContainerPeer");
                getInsetsMethod = Resolver.uncheck(Reflect.instanceFunction(containerPeerClass, "getInsets", Insets.class));
                Class<?> decoratedPeerClass = Class.forName("sun.awt.wl.WLDecoratedPeer");
                decorationField = Resolver.withAccess(decoratedPeerClass.getDeclaredField("decoration"));
                Class<?> frameDecorationClass = Class.forName("sun.awt.wl.WLFrameDecoration");
                markRepaintNeededMethod = Resolver.uncheck(Reflect.instanceProcedure(frameDecorationClass, "markRepaintNeeded"));
                isUndecoratedField = Resolver.withAccess(frameDecorationClass.getDeclaredField("isUndecorated"));
                Class<?> wlSurfaceClass = Class.forName("sun.awt.wl.WLMainSurface");
                getWlSurfacePtr = Resolver.uncheck(Reflect.instanceFunction(wlSurfaceClass, "getWlSurfacePtr", long.class));
            } catch (Throwable e) {
                throw new RuntimeException(e);
            }
        }

        public WL(Object peer) {
            this.peer = klazz.cast(peer);
        }

        public void performLocked(Runnable runnable) {
            performLockedMethod.accept(peer, runnable);
        }

        public void setDecorated(boolean decorated) {
            try {
                Object decoration = decorationField.get(peer);
                isUndecoratedField.setBoolean(decoration, !decorated);
                markRepaintNeededMethod.accept(decoration);
            } catch (IllegalAccessException e) {
                throw new RuntimeException(e);
            }
        }

        public boolean isDecorated() {
            return containerPeerClass.isInstance(peer) && getInsetsMethod.apply(peer).top > 0;
        }

        @Override
        public long nativePointer() {
            try {
                return nativePtrField.getLong(peer);
            } catch (IllegalAccessException e) {
                throw new RuntimeException(e);
            }
        }

        public long getSurfacePtr() {
            try {
                return getWlSurfacePtr.apply(wlSurfaceField.get(peer));
            } catch (IllegalAccessException e) {
                throw new RuntimeException(e);
            }
        }
    }

    class Resolver {
        private static final Logger logger = LoggerFactory.getLogger("dbusmenu4j/PeerResolver");
        private static final Field peerField;

        private static final MethodHandle accessibleSetter;

        static {
            try {
                peerField = Component.class.getDeclaredField("peer");
                peerField.setAccessible(true);
                accessibleSetter = CoreReflect.lookup(AccessibleObject.class).findSetter(AccessibleObject.class, "override", boolean.class);
            } catch (NoSuchFieldException | IllegalAccessException e) {
                logger.error("Failed to resolve peer field", e);
                throw new RuntimeException(e);
            }
        }

        private static Field withAccess(Field field) throws Throwable {
            accessibleSetter.invoke(field, true);
            return field;
        }

        private static Consumer<Object> uncheck(Consumer<?> in) {
            return (Consumer<Object>) in;
        }

        private static <R> Function<Object, R> uncheck(Function<?, R> in) {
            return (Function<Object, R>) in;
        }

        private static <T, R> BiFunction<Object, T, R> uncheck(BiFunction<?, T, R> in) {
            return (BiFunction<Object, T, R>) in;
        }

        private static <T> BiConsumer<Object, T> uncheck(BiConsumer<?, T> in) {
            return (BiConsumer<Object, T>) in;
        }

        public static Peer resolve(Component component) throws IllegalAccessException {
            return coerce(peerField.get(component));
        }
    }
}

File: src/main/kotlin/io/gitlab/jfronny/globalmenu/reflect/Peer.kt
package io.gitlab.jfronny.globalmenu.reflect

import io.gitlab.jfronny.commons.unsafe.reflect.Reflect
import java.awt.Component
import java.awt.Insets

sealed interface Peer {
    val nativePtr: Long

    class X11(private val inner: Any) : Peer {
        init {
            componentPeerClass.cast(inner)
        }
        override val nativePtr: Long get() = getPtrMethod(inner)

        companion object {
            val componentPeerClass = Class.forName("sun.awt.X11.XComponentPeer")
            private val getPtrMethod = Reflect.instanceFunction(componentPeerClass, "getWindow", Long::class.java).unchecked
        }
    }
    class WL(private val inner: Any) : Peer {
        init {
            componentPeerClass.cast(inner)
        }
        override val nativePtr: Long get() = nativePtrField.getLong(inner)
        fun performLocked(runnable: Runnable) {
            performLockedMethod(inner, runnable)
        }
        var decorated: Boolean
            get() = containerPeerClass.isInstance(inner) && getInsetsMethod(inner).top > 0
            set(value) {
                val decoration = decorationField.get(inner)
                isUndecoratedField.setBoolean(decoration, !value)
                markRepaintNeededMethod(decoration)
            }

        companion object {
            val componentPeerClass = Class.forName("sun.awt.wl.WLComponentPeer")
            private val performLockedMethod = Reflect.instanceProcedure(componentPeerClass, "performLocked", Runnable::class.java).unchecked1
            private val nativePtrField = componentPeerClass.getDeclaredField("nativePtr").withAccess
            private val containerPeerClass = Class.forName("java.awt.peer.ContainerPeer")
            private val getInsetsMethod = Reflect.instanceFunction(containerPeerClass, "getInsets", Insets::class.java).unchecked
            private val decoratedPeerClass = Class.forName("sun.awt.wl.WLDecoratedPeer")
            private val decorationField = decoratedPeerClass.getDeclaredField("decoration").withAccess
            private val frameDecorationClass = Class.forName("sun.awt.wl.WLFrameDecoration")
            private val markRepaintNeededMethod = Reflect.instanceProcedure(frameDecorationClass, "markRepaintNeeded").unchecked
            private val isUndecoratedField = frameDecorationClass.getDeclaredField("isUndecorated").withAccess
        }
    }

    companion object {
        operator fun invoke(peer: Any): Peer {
            if (peer.javaClass.name.contains("X11")) return X11(peer)
            if (peer.javaClass.name.contains("WL")) return WL(peer)
            throw IllegalArgumentException("Unknown peer type: ${peer.javaClass}")
        }
    }
}

private val peerField = Component::class.java.getDeclaredField("peer").apply { isAccessible = true }

val Component.peer: Peer get() = Peer(peerField.get(this))